### ISSUE TITLE
Problem with opengl

### ISSUE THREAD

**Issue by jgranick** [openfl/openfl-native#130]
#130 Issue by __delahee__,

@jgranick 
Hi stumbled on many problem accross the opengl API regarding uniformfv semantic.
It was clear the calls could not transmit the whole buffers.

Tried my best to fix here : 
https://github.com/motion-twin/openfl-native/commit/c1a0f067718e5fe7001f599200fc2b79427dfe4a
https://github.com/motion-twin/NME/commit/78255407fb51174e57a51cfc7d103acb4c81b4e8

Would be awesome if you could fix very quickly.

Additionnaly i couldnt test uniformMatrixFv calls and vertexAttribFv calls so you should setup and test them.

Thanks !

**Comment by jgranick** [comment]
Comment by __jgranick__:

Hi,

I'm a little unsure why this change is required.

The "uniform1fv" call, for example, has two signatures in WebGL:

```
void uniform1fv(WebGLUniformLocation? location, Float32Array v);
void uniform1fv(WebGLUniformLocation? location, sequence<GLfloat> v);
```

In either case, it is basically expecting an array of 32-bit float values. You should able to convert a standard Haxe array using "new Float32Array (myArray)", I believe, so either a standard array or a Float32Array object should be compatible. By accepting any IMemoryRange value, this means the API will accept 16-bit values, integers and other values which are not compatible with the API. 

Open to your input, but being stronger typed would be helpful

Thanks

**Comment by jgranick** [comment]
Comment by __delahee__:

You are both right and wrong. Previous code on hxcpp side was just not
working, the code was typed as Array<Float> which was an array of double
and then the array semantic was not respected, only first element was
bound, same for vattrib and uniform matrix, this has to be fixed :)

You are right in that it should only accept float ranges but i didnt find
the right call to do it because i am not familiar with byte access and
basically i just want to send a pointers to vram the type check was not my
priority :D

That is why i sent commits rather than pull reqs, i was feeling there was
room to enhancement here.

++

**Comment by jgranick** [comment]
Comment by __delahee__:

Reread your answer, you dont speak about the actual problem bit rather a
typing issue which is not the matter, so i will rephrase.

Read carefully.

The previous code sent double arrays which were truncated to the first
element. Then the shaders could not fetch values sent by user.

So the fix is about the user being able to send the values, the type
checking is of virtually no importance here.

So you have to fix the signature and find something better than mine in
order to guarantee the call intended by the user will provide what he wants
so that he can send a consistent buffer to gl.

This fix is critical for the  gl stack to work properly. Because witjout
that palette matrix and uniform vec arrays fail.

This shows stopped nicolas fully porting h3d.

Thanks for your diligency and love fixing this.

**Comment by jgranick** [comment]
Comment by __ukpaul__:

Very much looking forward to this one being fixed so I can see bone animation on native targets.

https://github.com/openfl/openfl-native/issues/20

**Comment by jgranick** [comment]
Comment by __delahee__:

I am on it dude :)

**Comment by jgranick** [comment]
Comment by __delahee__:

@jgranick 

Ok it is stable enought now, please merge : https://github.com/motion-twin/openfl-native/blob/master/openfl/gl/GL.hx and its corollary in motion-twin/openfl/gl/GL with lime/nme/openfl.

We have skinning and complex shaders on windows, android ans ios working.
